Handle HiDPI scaling in GEF diagrams

akoch-yatta opened the pull request:
I updated my approach for improving HiDPI scaling in GEF diagrams.

Central orchestrator is currently the `MonitorAwareZoomManager` which connects the `Control` (usually a `FigureCanvas`) and `ScalableFigures` to apply the monitor zoom as scale on. Linking is done via the `AbstractEditPartViewer` ( `MonitorAwareZoomManager` and `FigureCanvas`) `FreeformGraphicalRootEditPart/ScalableRootEditPart` (register `ScalableFigures`). Adaptions were made in `FreeformGraphicalRootEditPart` and `ScalableRootEditPart` to replace the innerLayer with a scalable layer that will registered into `MonitorAwareZoomManager`. For the flyout palette I added `SimpleAutoscaledRootEditPart` as a simple addition to do the same as the other two. The pattern is now everywhere one central scalable layer somewhere up in the hierarchy. 

It would be nice to move that into the root figure, but was not able to fully make it work. The problem described in https://github.com/eclipse-gef/gef-classic/pull/768#issuecomment-3168160441 by @ptziegler is to my analysis a "wrong" clipping caused by the zoom break in the root figure. Lets imaging having the flag active, so the FigureCanvas is at 100% and we have a monitor of 200%, so the root figure scale would be 2.0:
First, the client area of canvas is fetched, e.g. 1000x1000. Then the bounds of root figure are set with the value:

- if the bounds are scaled down, the bounds would be 500x500, which will be used as clipping area and will be too small for later calls
- if the bounds are not scaled down, so they remain 1000x1000, the clipping will be fine, but the children will be positioned at the wrong place

I was not able to properly solve this discrepancy, but I think that would enable us to move all the logic into a specific root figure.

One additional problem (for which I added some workarounds in the PR) are `TextUtilities` and `FigureUtilities` that use a static GC for its calculations. This means the GC will always the bound to the zoom that `new Shell()` is bounds to. At least in windows that mustn't even be the primary monitor, but could be a secondary or any additional monitor as well, which leads to cut off texts, when you wouldn't expect it. What we would need is the possibility to initialize the GC with a given control (e.g. either the parent Shell or the FigureCanvas) to provide the proper context for all calculations on this GC. As they are used at a lot of different places I didn't have a good idea, how to change these. This is a prerequisite for any proper solution for the underlying problem.

akoch-yatta committed:
Add autoscale disabled mode to allow drawing on unscaled GC

ptziegler commented:
> One additional problem (for which I added some workarounds in the PR) are TextUtilities and FigureUtilities that use a static GC for its calculations. This means the GC will always the bound to the zoom that new Shell() is bounds to. At least in windows that mustn't even be the primary monitor, but could be a secondary or any additional monitor as well, which leads to cut off texts, when you wouldn't expect it. 

I think you underestimate just how broken the `FigureUtilities` class is. It's also used together with the Printer classes, where the concept of a monitor doesn't exist in the first place. I think this is also what caused issues in e.g. https://github.com/eclipse-gef/gef-classic/pull/750.

Overall, I think the best approach is to slowly move away from those static methods and instead use a `GC` that is based on some local `Drawable` object.

I've added some suggestions of things where I think that they could be made a little bit cleaner. Overall, I think this is the most promising approach right now. I also played around with the root figure some more, but I never got it working the way I want.

Once you think that this is in a good state, I'm willing to give it a go. The release window just started, so we have more than enough time to test this in more detail and also handle any problems that might've been missed.

One thing I'd still like to see would be a decoupling of this zoom handling from GEF so that it can be re-purposed for Zest. But I hope that this is as simple as moving the `MonitorAwareZoomManager` to Draw2D.

ptziegler commented:
I tried to visualize my understanding of the current approach:

<img width="640" height="260" alt="hidpi-lws" src="https://github.com/user-attachments/assets/48e7b37d-97b8-435d-bf4f-9d5788b2e473" />

Instead of scaling the `RootFigure`, you went with scaling the `RootEditPart`. The zoom manager is bound to the `EditPartViewer` and is responsible for processing the device zoom of the hooked control. In order to receive those zoom changes, the edit parts need to register themselves to this manager.

ptziegler commented on org.eclipse.gef/src/org/eclipse/gef/internal/MonitorAwareZoomManager.java:
```diff
@@ -0,0 +1,85 @@
+import org.eclipse.draw2d.ScalableFigure;
+import org.eclipse.draw2d.internal.InternalDraw2dUtils;
+
+public class MonitorAwareZoomManager {
```
Maybe just as an alternative idea: The `EditPartViewer` supports `PropertyChangeListeners`. What if you forgo the `MonitorAwareZoomManager` and simply convert the `ZoomChangeEvent` into a `PropertyChangeEvent`?

The `RootEditPart` can extract the new zoom level from the event and set the scaling accordingly. This way you don't have to reinvent the wheel.

ptziegler replied:
Hm... That is indeed an issue. Though explicitly firing a `monitorScale` event when the `RootEditPart` has been swapped out still feels a little bit like a hack, but it's definitely an improvement.

I think conceptually, the "correct" approach would be to override the `activate()` method of the `RootEditPart`s and then initialize the scale with `getViewer().getProperty(MONITOR_SCALE_PROPERTY)`. What do you think?

ptziegler commented on org.eclipse.gef/src/org/eclipse/gef/internal/MonitorAwareZoomManager.java:
```diff
@@ -0,0 +1,85 @@
+	private double calculateScale() {
+		int zoom;
+		if (InternalDraw2dUtils.disableAutoscale) {
+			try {
+				zoom = (int) control.getData("NATIVE_ZOOM"); //$NON-NLS-1$
+			} catch (Exception e) {
+				zoom = 100;
+			}
```
Isn't the new scale factor the `detail` of the `ZoomChanged` event? Why not use that?

akoch-yatta commented on org.eclipse.gef/src/org/eclipse/gef/ui/palette/editparts/PaletteEditPart.java:
```diff
@@ -129,12 +135,71 @@ public Dimension getPreferredSize(int w, int h) {
+		}
+	}
+
+	public class PaletteTextUtilities extends TextUtilities {
```
No, of course not. I added all TextUtilities related changes just to make them somehow work. I wanted to wait for your feedback how to proceed with that.

azoitl commented on org.eclipse.gef/src/org/eclipse/gef/internal/MonitorAwareZoomManager.java:
```diff
@@ -0,0 +1,85 @@
+import org.eclipse.draw2d.ScalableFigure;
+import org.eclipse.draw2d.internal.InternalDraw2dUtils;
+
+public class MonitorAwareZoomManager {
```
As it is handling the native scaling of displays, how about NativeDisplayScaleManager?

ptziegler replied:
I think I mentioned it somewhere else already, but I don't think this class is needed, if the zoom changes are handled as PropertyChangeEvents.

akoch-yatta commented on org.eclipse.draw2d/src/org/eclipse/draw2d/FigureUtilities.java:
```diff
@@ -425,4 +429,9 @@ public static boolean isNotFullyClipped(IFigure figure) {
 		return !figBounds.isEmpty();
 	}
 
+	public static void configureAutoscaleMode(Control control) {
```
It doesn't always disable it, but only if the setting is set. I renamed it to configureForAutoscalingMode for now and moved it to InternalDraw2dUtils. But I am open to any better naming ideas.

akoch-yatta committed:
Manage disabled autoscaling for PopupHelper

If the disable autoscaling flag is set, e.g. tooltips would be scaled
to 100%. This commit scales and renders the Shell uses for PopupHelper
according to the monitor zoom the Shell is assigned to.

akoch-yatta commented on org.eclipse.draw2d/src/org/eclipse/draw2d/DrawableFigureUtilities.java:
```diff
@@ -0,0 +1,146 @@
+ *
+ * @since 3.27
+ */
+public class DrawableFigureUtilities {
```
I will add a separate commit to this PR soon for this problem. I have a solution that seems to work fine.

akoch-yatta committed:
Extract GC related operation in TextUtilities and FigureUtilities to be Drawable aware

This commit introduces two new utility classes to fully replace
TextUtilities and partially replace FigureUtilities, when calculations
on a GC are involded. Therefor DrawableTextUtilities or DrawableFigureUtilities
can to be instantiated with a Drawable passed as context when a GC is
instantiated.

akoch-yatta commented on org.eclipse.draw2d/src/org/eclipse/draw2d/PopUpHelper.java:
```diff
@@ -72,9 +74,8 @@ protected PopUpHelper(Control c, int shellStyle) {
-	@SuppressWarnings("static-method")
 	protected LightweightSystem createLightweightSystem() {
-		return new LightweightSystem();
+		return InternalDraw2dUtils.disableAutoscale ? new PopupHelperLightweightSystem() : new LightweightSystem();
```
I put that in the PopupHelper class for now. I don't know if you think it would be better to put it directly in the concrete two classes, but as it is only activated with the flag, I thought that would be fine.
The implementation of the special LightweightSystem is quite similar to the previous approaches and is lacking the same issues as them (problem with clipping), but I hope there are no use cases than run into these issues here, otherwise we would need to rethink

ptziegler replied:
I think putting it here is the better approach. Clients can use their own `PopupHelper` subclass, which would otherwise have no way to use this custom LWS. Now they can just override the `createLightweightSystem()` method if an issue appears and they need to revert to the previous implementation..

akoch-yatta commented:
@ptziegler I refactored and cleanup the commits up a bit. I hope you can still follow my latest changes.

azoitl commented:
@akoch-yatta @ptziegler thx for the the great work. I only found yesterday time to do a first look on this. I agree with both of you that we soon should get that merged for testing. 

I would try to test it over the weekend with 4diac IDE I hope that there I can see some special cases. For that I would have one question: In order that 4diac IDE is correctly drawing I need to know the height of my drawing font. So far I'm using the FigureUtilities for this. Can I still use this? Do I need my own GC with hardcoded 100% scaling now?

akoch-yatta commented on org.eclipse.draw2d/src/org/eclipse/draw2d/PopUpHelper.java:
```diff
@@ -219,4 +220,66 @@ protected void show() {
+				return c.getMonitor().getZoom() / 100.0;
+			} catch (NullPointerException e) {
+				return 1.0;
+			}
```
Yes, you are correct. that should never happen and I adapted it

akoch-yatta commented on org.eclipse.draw2d/src/org/eclipse/draw2d/ToolTipHelper.java:
```diff
@@ -121,7 +121,7 @@ public void displayToolTipNear(IFigure hoverSource, IFigure tip, int eventX, int
 			Dimension shellSize = getLightweightSystem().getRootFigure().getPreferredSize()
 					.getExpanded(getShellTrimSize());
-			setShellBounds(displayPoint.x, displayPoint.y, shellSize.width, shellSize.height);
+			setShellBounds(displayPoint.x, displayPoint.y, shellSize.width + 1, shellSize.height);
```
So, this issue has even a second layer, because scaling up and down will lead to the bounds of the label being set to 134 which is different to the preferred size of 135 and the label is adapted to use a truncated String with "..."

akoch-yatta replied:
I'm not a fan of this +1 solution as well, as it indicates there is a problem hidden underneath, so I prefer getting rid of the root cause as well

ptziegler commented on org.eclipse.draw2d/src/org/eclipse/draw2d/FigureUtilities.java:
```diff
@@ -73,7 +74,12 @@ public static FontMetrics getFontMetrics(Font f) {
+			InternalDraw2dUtils.configureForAutoscalingMode(shell);
+			gc = new GC(shell);
+			if (InternalDraw2dUtils.disableAutoscale) {
+				gc.setAdvanced(true);
```
I've got some mixed feelings about using advanced mode everywhere. Here is fine, of course, given that the GC should only be used internally. But if there is ever a need to create a GC somewhere else, it's incredibly easy to forget about it and you suddenly end up with inconsistent and hard to debug problems.

ptziegler commented:
@akoch-yatta How should we proceed with this one? I think the issue with the +1 for the width has been fixed by using the advanced mode in the `FigureUtilities`.

When it comes to how this should be handled in general, I would suggest a separate PR to keep each change isolated. Otherwise I fear that those discussion will drag on for a lot longer...

Can you also rebase the PR again to get rid of 89775da3bb7dec805672dad015d364dfd501f95e? What about the other commits. Do you want to keep them separate or should they be squashed.

Unless I've missed anything, this should then be ready to be comitted, right?

akoch-yatta commented:
@ptziegler I rebased the PR. I would propose to keep the three commits, because they tackle three different issues, but I can squash them if you want.

Overall I am in favor of merging this state to continuously improve it in the next weeks and get feedback in case we broke something - with and without setting the flag

azoitl reviewed (approved):
I tested during todays development with 4diac IDE under linux with wayland (100% & 125%).  Drawing wise all looks good. Used pallets, at least 6 different GEF Classic editors. 4diac IDE has a special figure pane which I was a bit nervouse. But everything worked out of the box. 

I had some issues with selection and hovering on 125%. But I think this is coming from SWT's wayland/gtk4 interaction. Because I also had it sometimes on editor tabs and menus.

Therefore I would go for merging it. 

@akoch-yatta thx a lot for this work.

@Phillipus this may be interesting for you as well.

akoch-yatta commented:
@ptziegler I did a small update to the PR. We noticed, that using the monitor zoom caused issues without monitor specific scaling active, therefor we switched back to using a data value from control. We will need the latest state from SWT so it contains commit https://github.com/eclipse-platform/eclipse.platform.swt/commit/648bb973ea19c646cc22bd9e3e71224b93a1fbdc for it to work.

ptziegler reviewed (approved):
I gave the latest state a try (outside of the new DATA_SHELL_ZOOM property) and things are looking really good. Even on Zest, the adaptations can be easily integrated.

@akoch-yatta Great job! Thanks for sticking to this to the end, even with all the constant back and forth.

feilimb commented:
Can I ask if there is any guide / wiki type document, which may assist developers of legacy custom GEF classic editors to adapt to recent changes & improvements with HiDPI scaling in Draw2d and the _MonitorAwareZoomManager_ etc.

After porting my tool to Eclipse 2026-03 we see issues in the areas of screen tearing, and connection lines misalignment with monitor at 150% Windows Scaling (eg. 4K monitor default setting).  I understand there is a _-Ddraw2d.enableAutoscale=false_ system property which can be set for legacy editors to use SWT scaling, but if we wish to pick up all the improvements in Draw2D scaling etc. and re-implement internal layout logic etc. it would be very useful if there was some guide on such a migration.

azoitl commented:
@feilimb unfortunately there is no such document. However I noticed from my work on 4diac IDE from the Eclipse 4diac project that we got position errors always when we where using Eclipse GEF Classic incorrectly. Not that this helps you. But to give you some pointers. I thing that we often did is to ask the zoom manager for the current zoom factor to convert between mouse coordinates to figure coordinates or from figure coordinates to connection layer/feedback layer. The correct way to do this is to use the `translateToAbsolute()` `translateToRelative()` methods for figures. 

If you have concrete examples of problems we are happy to support you.
